### DISCUSSION TITLE
jsonable_encoder exception on requests.Response object

### DISCUSSION THREAD

**Discussion by dg-ow** [fastapi/fastapi#8473] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import requests
from fastapi.encoders import jsonable_encoder

r = requests.get('https://api.github.com/thisodesntexist')

jsonable_encoder(r)
```

### Description

We have an API endpoint which returns a status object.  We had a status object which included a requests.Response object.  We received an exception from `jsonable_encoder`.  

I wrote a simple example script which simulates what happened.

```
$ pip3 list | grep requests
requests            2.25.1
requests-toolbelt   0.9.1
$ pip3 list | grep fastapi
fastapi             0.65.1
$ python3 ./encoder_exception.py 
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 134, in jsonable_encoder
    data = dict(obj)
TypeError: 'property' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 138, in jsonable_encoder
    data = vars(obj)
TypeError: vars() argument must have __dict__ attribute

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./encoder_exception.py", line 6, in <module>
    jsonable_encoder(r)
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 142, in jsonable_encoder
    return jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "/usr/local/lib/python3.8/site-packages/fastapi/encoders.py", line 141, in jsonable_encoder
    raise ValueError(errors)
ValueError: [TypeError("'property' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]
$ cat encoder_exception.py 
import requests
from fastapi.encoders import jsonable_encoder

r = requests.get('https://api.github.com/thisodesntexist')

jsonable_encoder(r)

```

### Operating System

Linux

### FastAPI Version

0.65.1

### Python Version

3.8.8

**Comment by Razikus** [comment, marked as answer]
Isn't it just beause requests.get is not jsonable?
You should return .json()

so
```
import requests
from fastapi.encoders import jsonable_encoder

r = requests.get('https://api.github.com/thisodesntexist')
jsoned = r.json()
jsonable_encoder(jsoned)
```

I don't know if this is an issue to encoder

**Comment by dg-ow** [comment]
My reading of the `jsonable_encoder` documentation is that it should return something that can be encoded in JSON, and doesn't say anything about limitations.

https://fastapi.tiangolo.com/tutorial/encoder/

**Comment by Razikus** [comment]
https://github.com/tiangolo/fastapi/blob/master/fastapi/encoders.py#L29

The limitations here are simple - Fastapi couldn't know about all objects and classes in the world
The idea about jsonable_encoder is that you should provide jsonable object

Response from Request isn't jsonable before you decide what type of response you want to have
https://docs.python-requests.org/en/latest/user/quickstart/#response-content

How Fastapi should interract with this?
Try to execute .json() command?

What if we will have another library that have toJson() command
Should it also try to execute toJson() method?

**Comment by dg-ow** [comment]
I see nothing in the code you linked me to documenting limitations.  I attempts to encode objects, with what looks like a last ditch failsafe at the bottom of the method.  The tutorial I linked to says the jsonable_encoder can return an JSON encodable object.  It even references a non-JSON encodable object, `datetime`.

So yes, I expect jsonable_encoder to give me back something that can be JSON encoded.  Maybe your idea of a toJson method is the way to go.

**Comment by tomy0000000** [comment]
If you looked into the [source code of `jsonable_encoder`](https://github.com/tiangolo/fastapi/blob/1876ebc77949a9a254909ec61ea0c09365169ec2/fastapi/encoders.py#L29), you'll find that it is basically an enhanced version of `json.dumps` with handling of `datetime`, `pydantic` models, `dataclass`, and some data structure that developer would "typically" return in the router function, no more and no less.

The error you saw, particularly
```
TypeError: 'property' object is not iterable
```
```
TypeError: vars() argument must have __dict__ attribute
```
these two, is actually `jsonable_encoder` try to guess how this object can be turned into json object

But the point is, it is not meant to support every type of object existed in python. Which is why you are the person responsible to call `r.json()` before returning it, since you know the object better than `jsonable_encoder`.

**Comment by grahamalama** [comment]
> My reading of the `jsonable_encoder` documentation is that it should return something that can be encoded in JSON, and doesn't say anything about limitations.
> 
> https://fastapi.tiangolo.com/tutorial/encoder/

> But the point is, it is not meant to support every type of object existed in python.

I wanted to chime in and say that I agree that the documentation here is a bit misleading. The documentation only says:

> [`jsonable_encoder`] receives an object, like a Pydantic model, and returns a JSON compatible version. 

It doesn't say:
> `jsonable_encoder` supports these object types `<...>`. If what you're trying to serialize isn't one of those types, it should have a `dict` method. Otherwise, `jsonable_encoder` will throw an error.

Or something along those lines. 

I _do_ get that it would be impossible to account for every type of Python object, but having those caveats in the documentation and maybe a more helpful error message would be useful.
